ext/standard/array.c

```diff
@@ -2057,13 +2055,25 @@ PHP_FUNCTION(array_fill_keys)
 }
 /* }}} */
 
-#define RANGE_CHECK_INIT_ARRAY(start, end) do { \
+#define RANGE_CHECK_DOUBLE_INIT_ARRAY(start, end) do { \
 		double __calc_size = ((start - end) / step) + 1; \
-		if (fabs(__calc_size) >= (double)HT_MAX_SIZE) { \
-			php_error_docref(NULL, E_WARNING, "The supplied range exceeds the maximum array size: start=%0.0f end=%0.0f", start > end ? end : start, start > end ? start : end); \
+		if (__calc_size >= (double)HT_MAX_SIZE) { \
+			php_error_docref(NULL, E_WARNING, "The supplied range exceeds the maximum array size: start=%0.0f end=%0.0f", end, start); \
+			RETURN_FALSE; \
+		} \
+		size = (uint32_t)__calc_size; \
+		array_init_size(return_value, size); \
+		zend_hash_real_init(Z_ARRVAL_P(return_value), 1); \
+	} while (0)
+
+#define RANGE_CHECK_LONG_INIT_ARRAY(start, end) do { \
+		double __calc_size = (start - end) / lstep; \
+		if (__calc_size >= HT_MAX_SIZE) { \
```

Comment:
You're right, `__calc_size` should be zend_ulong here. I think the check in the `RANGE_CHECK_LONG_INIT_ARRAY` is correct, and the one in `RANGE_CHECK_DOUBLE_INIT_ARRAY` is slightly off (it should only compare if `__calc_size > (double)HT_MAX_SIZE`, since 1 has already been added to account for the inclusive range).